tests/v2/validation/snapshot/snapshot_restore_k8s_upgrade_test.go

```diff
@@ -0,0 +1,108 @@
+//go:build (validation || extended) && !infra.any && !infra.aks && !infra.eks && !infra.gke && !infra.rke2k3s && !cluster.any && !cluster.custom && !cluster.nodedriver && !sanity && !stress
+
+package snapshot
+
+import (
+	"testing"
+
+	"github.com/rancher/rancher/tests/framework/clients/rancher"
+	"github.com/rancher/rancher/tests/framework/extensions/etcdsnapshot"
+
+	"github.com/rancher/rancher/tests/framework/pkg/config"
+	"github.com/rancher/rancher/tests/framework/pkg/session"
+
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+)
+
+type SnapshotRestoreK8sUpgradeTestSuite struct {
+	suite.Suite
+	session        *session.Session
+	client         *rancher.Client
+	clustersConfig *etcdsnapshot.Config
+}
+
+func (s *SnapshotRestoreK8sUpgradeTestSuite) TearDownSuite() {
+	s.session.Cleanup()
+}
+
+func (s *SnapshotRestoreK8sUpgradeTestSuite) SetupSuite() {
+	testSession := session.NewSession()
+	s.session = testSession
+
+	s.clustersConfig = new(etcdsnapshot.Config)
+	config.LoadConfig(etcdsnapshot.ConfigurationFileKey, s.clustersConfig)
+
+	client, err := rancher.NewClient("", testSession)
+	require.NoError(s.T(), err)
+
+	s.client = client
+}
+
+func (s *SnapshotRestoreK8sUpgradeTestSuite) TestRKE1SnapshotRestoreK8sUpgrade() {
+	snapshotRestoreK8sVersion := &etcdsnapshot.Config{
+		UpgradeKubernetesVersion: s.clustersConfig.UpgradeKubernetesVersion,
+		SnapshotRestore:          "kubernetesVersion",
+		IsRKE1:                   true,
+	}
+
+	snapshotRestoreAll := &etcdsnapshot.Config{
+		UpgradeKubernetesVersion: s.clustersConfig.UpgradeKubernetesVersion,
+		SnapshotRestore:          "all",
+		IsRKE1:                   true,
+	}
+
+	tests := []struct {
+		name         string
+		etcdSnapshot *etcdsnapshot.Config
+		client       *rancher.Client
+	}{
+		{"Restore Kubernetes version and etcd", snapshotRestoreK8sVersion, s.client},
+		{"Restore cluster config, Kubernetes version and etcd", snapshotRestoreAll, s.client},
+	}
+
+	for _, tt := range tests {
+		s.Run(tt.name, func() {
+			snapshotRestore(s.T(), s.client, s.client.RancherConfig.ClusterName, tt.etcdSnapshot, false)
+		})
+	}
+}
+
+func (s *SnapshotRestoreK8sUpgradeTestSuite) TestRKE2K3SSnapshotRestoreK8sUpgrade() {
+	snapshotRestoreK8sVersion := &etcdsnapshot.Config{
+		UpgradeKubernetesVersion: s.clustersConfig.UpgradeKubernetesVersion,
```

Comment:
I thought we agreed, that we would have default behavior where it would upgrade to the latest k8s version? Are we enforcing this now? I think I just need an explanation.

Reply:
We want to have the flexibility to decide if we are going to update to a specified K8s version or the default latest one. For both the static and the dynamic test, this flexibility is achieved because if you have the K8s version blank in your config file, the code in `snapshots.go` will use the latest K8s version.

If it's not empty, then it will upgrade to your specified K8s version.

Reply:
Ah ok I missed that. I just needed an explanation. Thank you!